group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -750,7 +770,9 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
 
         // Get or create the member.
         if (memberId.isEmpty()) memberId = Uuid.randomUuid().toString();
-        final ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        final ConsumerGroupMember member = instanceId == null ?
+                group.getOrMaybeCreateMember(memberId, createIfNotExists) :
+                group.getOrMaybeCreateStaticMember(memberId, instanceId, createIfNotExists);
         throwIfMemberEpochIsInvalid(member, memberEpoch, ownedTopicPartitions);
 
         if (memberEpoch == 0) {
```

Comment:
I wonder if we could log something here as well when a static member is replaced.

Reply:
I just thought about something else. When a static member is replaced, we need to write records to erase the state of the previous member.